pkg/app/ops/deploymentchaincontroller/controller.go

```diff
@@ -0,0 +1,177 @@
+		deploymentStore:      datastore.NewDeploymentStore(ds),
+		deploymentChainStore: datastore.NewDeploymentChainStore(ds),
+		updaters:             make(map[string]*updater),
+		logger:               logger.Named("deployment-chain-controller"),
+	}
+}
+
+func (d *DeploymentChainController) Run(ctx context.Context) error {
+	syncUpdatersTicker := time.NewTicker(syncUpdatersInterval)
+	defer syncUpdatersTicker.Stop()
+	syncDeploymentChainsTicker := time.NewTicker(syncDeploymentChainsInterval)
+	defer syncDeploymentChainsTicker.Stop()
+
+	d.logger.Info("start running deployment chain controller")
+	defer d.logger.Info("deployment chain controller has been stopped")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-syncUpdatersTicker.C:
+			if err := d.syncUpdaters(ctx); err != nil {
+				d.logger.Error("failed while sync controller updaters", zap.Error(err))
+			}
+		case <-syncDeploymentChainsTicker.C:
+			if err := d.syncDeploymentChains(ctx); err != nil {
+				d.logger.Error("failed while sync deployment chains", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (d *DeploymentChainController) syncUpdaters(ctx context.Context) error {
+	// Remove done updater of completed deployment chain.
+	for id, u := range d.updaters {
+		if !u.IsDone() {
+			continue
+		}
+		d.logger.Info("remove done updater of deployment chain",
+			zap.String("id", id),
+			zap.Time("completed_at", u.doneTimestamp),
+		)
+		delete(d.updaters, id)
+	}
+
+	// Find all not completed deployment chains and create updater if does not exist.
+	notCompletedChains, err := listNotCompletedDeploymentChain(ctx, d.deploymentChainStore)
+	if err != nil {
+		d.logger.Error("failed to fetch all not completed deployment chain", zap.Error(err))
+		return err
+	}
+	for _, c := range notCompletedChains {
+		// Ignore in case there is updater for that deployment chain existed.
+		if _, ok := d.updaters[c.Id]; ok {
+			continue
+		}
+		d.updaters[c.Id] = newUpdater(
+			c,
+			d.applicationStore,
+			d.deploymentStore,
+			d.deploymentChainStore,
+			d.logger,
+		)
+	}
+
+	return nil
+}
+
+func (d *DeploymentChainController) syncDeploymentChains(ctx context.Context) error {
+	var (
+		updatersNum = len(d.updaters)
+		updaterCh   = make(chan *updater, updatersNum)
+		wg          sync.WaitGroup
+	)
+	updaterWorkerNum := maxUpdaterWorkerNum
+	if updaterWorkerNum > updatersNum {
+		updaterWorkerNum = updatersNum
+	}
+
+	d.logger.Info(fmt.Sprintf("there are %d running deployment chain updaters", updatersNum))
+	for w := 0; w < updaterWorkerNum; w++ {
+		wg.Add(1)
+		go func(wid int) {
+			d.logger.Info(fmt.Sprintf("worker id (%d) is handling deployment chain updaters", wid))
+			defer wg.Done()
+
+			for updater := range updaterCh {
+				updater.Run(ctx)
+			}
+			d.logger.Info(fmt.Sprintf("worker id (%d) has stopped", wid))
+		}(w)
+	}
+
+	for chainID := range d.updaters {
+		// Ignore updaters which be marked as done.
+		if d.updaters[chainID].IsDone() {
+			continue
+		}
```

Comment:
This check should be done inside `updater.Run` to be simple.
Because the number of goroutines (line 131) is determined based on the original list (including "done" ones).

Reply:
Ah, sure, this one is remained from my wrong lock stub 👍

Reply:
addressed by https://github.com/pipe-cd/pipe/pull/2986/commits/cda7818aebea96719b3635523f513d5e64d6b9df